### PR TITLE
[Feature] Support TAO dataset of MOT field

### PR DESCRIPTION
**Pull request by Seerkfang**

### CODE REVIEW

File: mmtrack/datasets/tao_dataset.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                # naming convention of 000000000000.jpg
+                # (LVIS v1 will fix this naming issue)
+                info['filename'] = info['file_name'][-16:]
+            else:
+                info['filename'] = info['file_name']
+            data_infos.append(info)
+        return data_infos
+
+    def load_tao_anns(self, ann_file):
+        self.coco = CocoVID(ann_file)
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:59]: docstring

File: mmtrack/datasets/tao_dataset.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        for _img_infos, _results in zip(img_infos, results):
+            track_ids = []
+            for img_info, result in zip(_img_infos, _results):
+                img_id = img_info['id']
+                for label in range(len(result)):
+                    bboxes = result[label]
+                    for i in range(bboxes.shape[0]):
+                        data = dict(
+                            image_id=img_id,
+                            bbox=self.xyxy2xywh(bboxes[i, 1:]),
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:106]: `self.xyxy2xywh` no exist, use the same api from mmdet

File: mmtrack/datasets/tao_dataset.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+                    json_results.append(data)
+        return json_results
+
+    def format_results(self, results, resfile_path=None):
+        """Format the results to json (standard format for COCO evaluation).
+
+        Args:
+            results (list[tuple | numpy.ndarray]): Testing results of the
+                dataset.
+            jsonfile_prefix (str | None): The prefix of json files. It includes
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:143]: `resfile_path`

File: mmtrack/datasets/tao_dataset.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        for metric in metrics:
+            if metric not in allowed_metrics:
+                raise KeyError(f'metric {metric} is not supported.')
+
+        result_files, tmp_dir = self.format_results(results, resfile_path)
+
+        eval_results = dict()
+
+        if 'track' in metrics:
+            from tao.toolkit.tao import TaoEval
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:195]: put it in the front of the file, and use `try...except taoeval=None` to import the module

File: mmtrack/datasets/tao_dataset.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            tao_eval.print_results()
+            tao_results = tao_eval.get_results()
+            for k, v in tao_results.items():
+                if isinstance(k, str) and k.startswith('AP'):
+                    key = 'track_{}'.format(k)
+                    val = float('{:.3f}'.format(float(v)))
+                    eval_results[key] = val
+
+        if 'bbox' in metrics:
+            from lvis import LVIS, LVISEval, LVISResults
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:212]: the same as above

File: tests/test_data/test_datasets/test_tao_dataset.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import os.path as osp
+
+from mmtrack.datasets import DATASETS as DATASETS
+from .utils import _create_coco_gt_results
+
+PREFIX = osp.join(osp.dirname(__file__), '../../data')
+DEMO_ANN_FILE = f'{PREFIX}/demo_tao_data/ann.json'
+DEMO_CLASSES_FILE = f'{PREFIX}/demo_tao_data/tao_classes.txt'
```
Comment (GT9505) [inline tests/test_data/test_datasets/test_tao_dataset.py:8]: delete the txt.
we can pass a list, ranther than a txt file

File: tools/convert_datasets/tao/merge_coco_with_lvis.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+from tqdm import tqdm
+
+
+def main():
+    # Use first line of file docstring as description if it exists.
+    parser = argparse.ArgumentParser(
+        description=__doc__.split('\n')[0] if __doc__ else '',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument(
+        '--lvis', type=Path, requirged=True, help='lvis json path')
```
Comment (GT9505) [inline tools/convert_datasets/tao/merge_coco_with_lvis.py:20]: `requirged` typo
you need to pull origin master to use the newest precommit config

File: tools/convert_datasets/tao/merge_coco_with_lvis.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    parser.add_argument(
+        '--iou-thresh',
+        default=0.7,
+        type=float,
+        help=('If a COCO annotation overlaps with an LVIS annotations with '
+              'IoU over this threshold, we use only the LVIS annotation.'))
+
+    args = parser.parse_args()
+    args.output.parent.mkdir(exist_ok=True, parents=True)
+    common_setup(args.output.name + '.log', args.output.parent, args)
```
Comment (GT9505) [inline tools/convert_datasets/tao/merge_coco_with_lvis.py:38]: why use this func?

File: tools/convert_datasets/tao/merge_coco_with_lvis.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    synset2lvis = {cat['syn set']: cat['id'] for cat in lvis.cats.values()}
+    coco2lvis = {}
+    with open(args.mapping, 'r') as f:
+        mapping = json.load(f)
+    for cat in coco.cats.values():
+        mapped = mapping[cat['name']]
+        assert mapped['coco_cat_id'] == cat['id']
+        synset = mapped['synset']
+        if synset not in synset2lvis:
+            logging.debug(
```
Comment (GT9505) [inline tools/convert_datasets/tao/merge_coco_with_lvis.py:58]: We don't need a log file to record the ouput information
Just print it

File: tools/convert_datasets/tao/tao2coco.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+    for file in [
+            'train.json', 'validation.json', 'test_without_annotations.json'
+    ]:
+        print(f'convert {file}')
+        out = convert_tao(osp.join(args.tao, file), classes)
+        c = '_482' if args.filter_classes else ''
+        prefix = file.split('.')[0].split('_')[0]
+        out_file = f'{prefix}{c}_ours.json'
+        mmcv.dump(out, osp.join(args.tao, out_file))
```
Comment (GT9505) [inline tools/convert_datasets/tao/tao2coco.py:154]: change the saved filename to `train_482classes`
i.e., you need to change to `out_file = f'{prefix}{c}classes.json'`

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -52,6 +52,8 @@ jobs:
         run: |
           pip install git+https://github.com/votchallenge/toolkit.git
           pip install attributee==0.1.5
+          pip install git+https://github.com/TAO-Dataset/tao.git
+          pip install git+https://github.com/lvis-dataset/lvis-api.git
```
Comment (GT9505) [inline .github/workflows/build.yml:56]: change CI. Move them below

File: docs/en/dataset.md [hunk 12]
```diff
@@ -29,7 +30,13 @@ Please download the datasets from the official websites. It is recommended to sy
...
 #### 1.2 Multiple Object Tracking
 
-- For the training and testing of multi object tracking task, one of the MOT Challenge datasets (e.g. MOT17) is needed, and CrowdHuman can be served as comlementary dataset.
+- For the training and testing of multi object tracking task, one of the MOT Challenge datasets (e.g. MOT17) and TAO are needed, and CrowdHuman can be served as comlementary dataset.
+
+- For TAO dataset, you need to download annotations from the official website manually, and move the folder to the subfolder of tao as the following data structure shows.
+
+```shell
+git clone https://github.com/TAO-Dataset/annotations.git
+```
```
Comment (GT9505) [inline docs/en/dataset.md:39]: Why do you need `annotations.git`

File: docs/en/dataset.md [hunk 13]
```diff
@@ -288,6 +316,26 @@ mmtracking
 │   │   │   ├── crowdhuman_train.json
 │   │   │   ├── crowdhuman_val.json
 │   │
+│   ├── tao
+│   │   ├── annotations
+│   │   │   ├── test_482.json
```
Comment (GT9505) [inline docs/en/dataset.md:321]: I remember to change `test_482.json` to `test_482_classes.json`

File: docs/en/install.md [hunk 14]
```diff
@@ -125,6 +125,13 @@ The compatible MMTracking, MMCV, and MMDetection versions are as below. Please i
     Detailed information of this disadaptation can be found at the following link:
     "https://github.com/votchallenge/toolkit/issues/42#issuecomment-1033267539"
 
+8. Install extra dependencies for TAO evaluation
+
```
Comment (GT9505) [inline docs/en/install.md:129]: merge it with MOT chanllange

File: docs/en/dataset.md [hunk 15]
```diff
@@ -197,6 +217,10 @@ python ./tools/convert_datasets/mot/mot2reid.py -i ./data/MOT17/ -o ./data/MOT17
 # CrowdHuman
 python ./tools/convert_datasets/mot/crowdhuman2coco.py -i ./data/crowdhuman -o ./data/crowdhuman/annotations
 
+# TAO
+# Generate filtered json file for QDTrack
+python ./tools/convert_datasets/tao/tao2coco.py -i ./data/tao/annotations --filter-classes
```
Comment (GT9505) [inline docs/en/dataset.md:222]: Add lvis dataset
Add the script of merging lvis and coco dataset

File: docs/en/install.md [hunk 16]
```diff
@@ -125,6 +125,13 @@ The compatible MMTracking, MMCV, and MMDetection versions are as below. Please i
     pip install git+https://github.com/JonathonLuiten/TrackEval.git
     ```
 
+9. Install extra dependencies for TAO evaluation.
```
Comment (GT9505) [inline docs/en/install.md:128]: please merge 8 and 9:
`8. Install extra dependencies:`
`- For MOTChallenge evaluation:`
`...`
`- For TAO evaluation:`
`...`

File: tests/data/demo_tao_data/tao_classes.txt [hunk 17]
```diff
@@ -0,0 +1,2 @@
+serving_dish
```
Comment (GT9505) [inline tests/data/demo_tao_data/tao_classes.txt:1]: delete this file

File: docs/en/dataset.md [hunk 18]
```diff
@@ -469,6 +513,20 @@ There are 2 JSON files in `data/crowdhuman/annotations`:
 `crowdhuman_train.json`:  JSON file containing the annotations information of the training set in CrowdHuman dataset.
 `crowdhuman_val.json`:  JSON file containing the annotations information of the validation set in CrowdHuman dataset.
 
+### The folder of annotations in tao
+
+There are 9 JSON files in `data/tao/annotations`:
+
+`test_categories.json`: JSON file containing a list of categories which will be evaluated on the TAO test set.
+`test_without_annotations.json`:  JSON for test videos. The 'images' and 'videos' fields contain the images and videos that will be evaluated on the test set.
+`test_482.json`: JSON file containing the converted results for test set.
```
Comment (GT9505) [inline docs/en/dataset.md:522]: forgot to change `test_482.json` to `test_482_classes.json`

File: mmtrack/datasets/tao_dataset.py [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+                for label in range(len(result)):
+                    bboxes = result[label]
+                    for i in range(bboxes.shape[0]):
+                        data = dict(
+                            image_id=img_id,
+                            bbox=self.xyxy2xywh(bboxes[i, 1:]),
+                            score=float(bboxes[i][-1]),
+                            video_id=img_info['video_id'],
+                            category_id=label
+                            if len(results) == 1230 else self.cat_ids[label],
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:130]: add comment to explain that `1230` denotes the number of categories in LVIS dataset.

File: mmtrack/datasets/tao_dataset.py [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+            result = results[idx]
+            for label in range(len(result)):
+                bboxes = result[label]
+                for i in range(bboxes.shape[0]):
+                    data = dict(
+                        image_id=img_id,
+                        bbox=self.xyxy2xywh(bboxes[i]),
+                        score=float(bboxes[i][4]),
+                        category_id=label
+                        if len(result) == 1230 else self.cat_ids[label])
```
Comment (GT9505) [inline mmtrack/datasets/tao_dataset.py:153]: the same as above

File: docs/en/dataset.md [hunk 21]
```diff
@@ -107,6 +110,12 @@ mmtracking
 │   │   │   ├── Images
 │   │   │   ├── CrowdHuman_val.zip
 │   │
+│   ├── lvis
+│   │   ├── train (the same as coco/train2017)
+│   │   ├── val (the same as coco/val2017)
+│   │   ├── test (the same as coco/test2017)
+│   │   ├── annotations
```
Comment (GT9505) [inline docs/en/dataset.md:117]: please list the filenames under `annotataions\`

File: docs/en/dataset.md [hunk 22]
```diff
@@ -30,10 +31,12 @@ Please download the datasets from the official websites. It is recommended to sy
 
 #### 1.2 Multiple Object Tracking
 
-- For the training and testing of multi object tracking task, one of the MOT Challenge datasets (e.g. MOT17) and TAO are needed, and CrowdHuman can be served as comlementary dataset.
+- For the training and testing of multi object tracking task, one of the MOT Challenge datasets (e.g. MOT17) and TAO are needed, CrowdHuman and LVIS can be served as comlementary dataset.
 
 - The `annotations` under `tao` contains the official annotations from [here](https://github.com/TAO-Dataset/annotations).
 
+- The `annotations` under `lvis` contains the official annotations can be downloaded according to [here](https://github.com/lvis-dataset/lvis-api/issues/23#issuecomment-894963957)。
```
Comment (GT9505) [inline docs/en/dataset.md:38]:
```suggestion
The `annotations` under `lvis` contains the official annotations of lvis-v0.5 which can be downloaded according to [here](https://github.com/lvis-dataset/lvis-api/issues/23#issuecomment-894963957)。
```

File: docs/en/dataset.md [hunk 23]
```diff
@@ -217,6 +226,10 @@ python ./tools/convert_datasets/mot/mot2reid.py -i ./data/MOT17/ -o ./data/MOT17
 # CrowdHuman
 python ./tools/convert_datasets/mot/crowdhuman2coco.py -i ./data/crowdhuman -o ./data/crowdhuman/annotations
 
+# LVIS
+# Merge annotations from LVIS and COCO for training QDTrack
+python ./tools/convert_datasets/tao/merge_coco_with_lvis.py --lvis ./data/lvis/annotations/lvis_v0.5_train.json --coco ./data/coco/annotations/instances_train2017.json --mapping ./data/coco_to_lvis_synset.json --output-json ./data/lvis/annotations/lvisv0.5+coco_train.json
```
Comment (GT9505) [inline docs/en/dataset.md:231]: please also tell users where to download `coco_to_lvis_synset.json`

File: docs/en/dataset.md [hunk 24]
```diff
@@ -312,6 +325,13 @@ mmtracking
 │   │   │   ├── crowdhuman_train.json
 │   │   │   ├── crowdhuman_val.json
 │   │
+│   ├── lvis
+│   │   ├── train (the same as coco/train2017)
+│   │   ├── val (the same as coco/val2017)
+│   │   ├── test (the same as coco/test2017)
+│   │   ├── annotations
+│   │   │   ├── lvisv0.5+coco_train.json
```
Comment (GT9505) [inline docs/en/dataset.md:333]: list the filenames under `annotations`

File: docs/en/dataset.md [hunk 25]
```diff
@@ -29,7 +31,13 @@ Please download the datasets from the official websites. It is recommended to sy
...
 #### 1.2 Multiple Object Tracking
 
-- For the training and testing of multi object tracking task, one of the MOT Challenge datasets (e.g. MOT17) is needed, and CrowdHuman can be served as comlementary dataset.
+- For the training and testing of multi object tracking task, one of the MOT Challenge datasets (e.g. MOT17) and TAO are needed, CrowdHuman and LVIS can be served as comlementary dataset.
+
+- The `annotations` under `tao` contains the official annotations from [here](https://github.com/TAO-Dataset/annotations).
+
+- The `annotations` under `lvis` contains the official annotations of lvis-v0.5 which can be downloaded according to [here](https://github.com/lvis-dataset/lvis-api/issues/23#issuecomment-894963957)。
+
+- The synset mapping file `coco_to_lvis_synset.json` used in `lvis` training can be found [here](https://github.com/lvis-dataset/lvis-api/issues/23#issuecomment-894963957)。
```
Comment (GT9505) [inline docs/en/dataset.md:40]: wrong link.

File: docs/en/dataset.md [hunk 26]
```diff
@@ -197,6 +234,14 @@ python ./tools/convert_datasets/mot/mot2reid.py -i ./data/MOT17/ -o ./data/MOT17
 # CrowdHuman
 python ./tools/convert_datasets/mot/crowdhuman2coco.py -i ./data/crowdhuman -o ./data/crowdhuman/annotations
 
+# LVIS
+# Merge annotations from LVIS and COCO for training QDTrack
+python ./tools/convert_datasets/tao/merge_coco_with_lvis.py --lvis ./data/lvis/annotations/lvis_v0.5_train.json --coco ./data/coco/annotations/instances_train2017.json --mapping ./data/coco_to_lvis_synset.json --output-json ./data/lvis/annotations/lvisv0.5+coco_train.json
```
Comment (GT9505) [inline docs/en/dataset.md:239]: put `./data/coco_to_lvis_synset.json` into `./data/lvis/annotations/coco_to_lvis_synset.json`

File: docs/en/dataset.md (hunk 25 above)
Comment (GT9505) [reply docs/en/dataset.md:40]: And merge line 38 with line 40

**Review by GT9505 — APPROVED** [review]
